### PR TITLE
Add handling ChannelVisibleEvent

### LINKED ISSUE #3510: ChannelListView is not refreshed correctly after unhiding the channel
### Context & proposal

The unhidden channel should be visible in the list.

### PR DESCRIPTION
**Pull request by adasiewiczr**
closes #3510 

### 🎯 Goal

Add handling `ChannelVisibleEvent`.

### 🛠 Implementation details

1. Added new `EventHandlingResult.WatchAndAdd`. After returning this result the SDK will call watchChannel and add it to the channels list
2. Added handling `ChannelVisibleEvent` in `BaseEventHandler`. By default it returns `EventHandlingResult.WatchAndAdd`

### 🎨 UI Changes

<table>
<thead>
<tr>
<th>Before</th>
<th>After</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<video src="https://user-images.githubusercontent.com/17440581/169814762-10fe1a9c-8935-48ff-8e64-c231af90f9c8.mp4" controls="controls" muted="muted" />
</td>
<td>
<video src="https://user-images.githubusercontent.com/17440581/169814751-90f8f316-362b-4929-b03c-d451ba0005bd.mp4" controls="controls" muted="muted" />
</td>
</tr>
</tbody>
</table>

### 🧪 Testing
Apply the patch. It changes behavior of ChannelListHeaderView's buttons. The left one (user avatar) hides `ChannelToHide` and the right one shows the same channel:
<details>
  <summary>Patch file to apply these changes for testing</summary>

```
Index: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/list/header/ChannelListHeaderView.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/list/header/ChannelListHeaderView.kt b/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/list/header/ChannelListHeaderView.kt
--- a/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/list/header/ChannelListHeaderView.kt	(revision 5b2d32a268d6187c4f9f4e15d1fdbd9659bae920)
+++ b/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/list/header/ChannelListHeaderView.kt	(date 1653307296346)
@@ -25,6 +25,8 @@
 import androidx.core.content.res.use
 import androidx.core.view.isInvisible
 import androidx.core.view.isVisible
+import io.getstream.chat.android.client.ChatClient
+import io.getstream.chat.android.client.logger.ChatLogger
 import io.getstream.chat.android.client.models.User
 import io.getstream.chat.android.core.internal.InternalStreamChatApi
 import io.getstream.chat.android.ui.R
@@ -234,14 +236,24 @@
      * the current user.
      */
     public fun setOnUserAvatarClickListener(listener: UserAvatarClickListener) {
-        binding.userAvatar.setOnClickListener { listener.onUserAvatarClick() }
+        binding.userAvatar.setOnClickListener {
+            // listener.onUserAvatarClick()
+            ChatClient.instance().channel("messaging:fe787dfe-0cd7-4ca9-99b5-3b40c9347d66").hide().enqueue {
+                ChatLogger.get("RafalTest").logE("Hiding successful: ${it.isSuccess}")
+            }
+        }
     }
 
     /**
      * Sets a click listener for the right button in the header.
      */
     public fun setOnActionButtonClickListener(listener: ActionButtonClickListener) {
-        binding.actionButton.setOnClickListener { listener.onClick() }
+        binding.actionButton.setOnClickListener {
+            // listener.onClick()
+            ChatClient.instance().channel("messaging:fe787dfe-0cd7-4ca9-99b5-3b40c9347d66").show().enqueue {
+                ChatLogger.get("RafalTest").logE("Showing successful: ${it.isSuccess}")
+            }
+        }
     }
 
     @InternalStreamChatApi

```
</details>

1. Sign in as Jc and verify `ChannelToHide` is visible. Click on the right button if it's not
2. Click on user avatar - channel should be hidden
3. Click on left button - channel should be visible

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [x] New code is covered by unit tests
- [x] Comparison screenshots added for visual changes
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### CODE REVIEW

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/logic/querychannels/internal/QueryChannelsLogic.kt [hunk 1]
```diff
@@ -125,11 +125,26 @@ internal class QueryChannelsLogic(
...
+     * @param cid cid of the channel.
+     */
+    private suspend fun watchAndAddChannel(cid: String) {
+        val result = client.channel(cid = cid).watch().await()
+
+        logger.logE(result.isSuccess.toString())
+        if (result.isSuccess) {
+            logger.logE(result.data().toString())
+            addChannel(result.data())
+        }
```
Comment (kanat) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/logic/querychannels/internal/QueryChannelsLogic.kt:145]: Hey @adasiewiczr,

Should we log the success result with an error message?

**Review by amitkma — APPROVED** [review]
LGTM
### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

**Review by kanat — APPROVED** [review]
LGTM

### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

**Review by TGazica — APPROVED** [review]
Looks good to me.

### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/logic/querychannels/internal/QueryChannelsLogic.kt (hunk 1 above)
Comment (adasiewiczr) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/logic/querychannels/internal/QueryChannelsLogic.kt:145]: good catch!

**Pull request merged**